Optimize HTMLConstructionSite::mergeAttributesFromTokenIntoElement()

cdumez opened the pull request:
#### 545d30c935cc3042ae868841e0c2789bc291bf26
<pre>
Optimize HTMLConstructionSite::mergeAttributesFromTokenIntoElement()
<a href="https://bugs.webkit.org/show_bug.cgi?id=244604">https://bugs.webkit.org/show_bug.cgi?id=244604</a>

Reviewed by Darin Adler.

Avoid duplicate calls to findAttributeByName() for each attribute both inside
mergeAttributesFromTokenIntoElement() and Element::setAttribute(). Instead,
introduce a new Element::setAttributeWithoutOverwriting() and drop now
unnecessary call to findAttributeByName() at call site.

This is a 0.6% progression on Speedometer on Intel and seems to be a 0.3%
progression on Apple Silicon.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setAttributeWithoutOverwriting):
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::mergeAttributesFromTokenIntoElement):

Canonical link: <a href="https://commits.webkit.org/254046@main">https://commits.webkit.org/254046@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/732f014f0b44f1025285f386097ad8bc350d58ac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87892 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32003 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18603 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97057 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152021 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91858 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/64/builds/30379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26384 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79982 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91800 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93505 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/64/builds/30379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/74590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/24479 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/64/builds/30379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/28101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/28197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14425 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2845 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/31221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/74590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/30171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33700 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
I looked for additional things to tighten up but didn’t find any major ones. One possibility might be to call `ensureElementData()` one time outside the loop if the attributes array is not empty and call versions of the functions that don’t have `elementData()` null checks.

weinig commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1874,6 +1874,17 @@ void Element::setAttribute(const QualifiedName& name, const AtomString& value)
     setAttributeInternal(index, name, value, NotInSynchronizationOfLazyAttribute);
 }
 
+void Element::mergeAttributes(const Vector<Attribute>& attributes)
```
Can this be pass by value to allow callers to move the vector in? In that case, you could probably avoid some ref/derefs by moving the attributes into addAttributeInternal

weinig commented on Source/WebCore/dom/Element.h:
```diff
@@ -115,6 +115,8 @@ class Element : public ContainerNode {
     template<typename... QualifiedNames>
     inline const AtomString& getAttribute(const QualifiedName&, const QualifiedNames&...) const;
     WEBCORE_EXPORT void setAttribute(const QualifiedName&, const AtomString& value);
+    void setAttributeWithoutOverwriting(const QualifiedName&, const AtomString& value);
```
Is this used?

cdumez committed:
Optimize HTMLConstructionSite::mergeAttributesFromTokenIntoElement()
https://bugs.webkit.org/show_bug.cgi?id=244604

Reviewed by Darin Adler.

Avoid duplicate calls to findAttributeByName() for each attribute both inside
mergeAttributesFromTokenIntoElement() and Element::setAttribute(). Instead,
introduce a new Element::setAttributeWithoutOverwriting() and drop now
unnecessary call to findAttributeByName() at call site.

This is a 0.6% progression on Speedometer on Intel and seems to be a 0.3%
progression on Apple Silicon.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setAttributeWithoutOverwriting):
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::mergeAttributesFromTokenIntoElement):

Canonical link: https://commits.webkit.org/254046@main
